src/algorithms/dqns/common.jl

```diff
@@ -35,3 +35,42 @@ function RLBase.update!(
     push!(trajectory[:terminal], is_terminated(env))
     push!(trajectory[:priority], p.learner.default_priority)
 end
+
+#####
+# some common components for Dueling network
+#####
+
+struct DuelingNetwork
+    base::Chain
+    val::Chain
+    adv::Chain
+end
+
+function build_dueling_network(network::Chain, device, n_atoms::Int=1)
+    lm = length(network)
+    if !(network[lm] isa Dense) || !(network[lm-1] isa Dense) 
+        error("The qnetwork provided is incompatible with dueling.")
+    end
+    base = Chain([deepcopy(network[i]) for i=1:lm-2]...) |> device
+    last_layer_dims = size(network[lm].W, 2)
+    val = Chain(deepcopy(network[lm-1]), Dense(last_layer_dims, n_atoms)) |> device
+    adv = Chain([deepcopy(network[i]) for i=lm-1:lm]...) |> device
+    return DuelingNetwork(base, val, adv)
+end
+
+Flux.@functor DuelingNetwork
+
+function (m::DuelingNetwork)(state)
+    x = m.base(state)
+    val = m.val(x)
+    if size(val)[1] == 1
+        # DQN Case
+        return val .+ m.adv(x) .- mean(m.adv(x), dims=1)
```

Comment:
I understand that you want the `DuelingNetwork` to be more flexible here. However, I prefer to only implement the most common case here. And implement the special distributional case in Rainbow or any other places.

Reply:
That is, we just need to implement dueling network in Rainbow?

Reply:
![image](https://user-images.githubusercontent.com/5612003/114817667-04e72800-9ded-11eb-9773-a8c45bb0c795.png)

I mean that the original Dueling Network doesn't consider the distributional case. So we just need to define **the most common implementation** here. And when applying the dueling network in Rainbow, we can add the specific part there. That's why I ask you to  to move `build_dueling_network` outside of `common.jl`.

Reply:
Do note that, just like `GaussianNetwork`, `DuelingNetwork` is a common network structure. It may be applied to DQN, Rainbow or some other algorithms. That's why just need to implement the most basic parts around `DuelingNetwork`.

Reply:
In which file should I implement the Dueling Network? Create a new .jl file?

Reply:
It's fine to keep it in `dqns/common.jl` currently. But in the next minor release, we should move it and `GaussianNetwork` into `RLCore`.

Reply:
If we don't design an if branch, how can we distinguish between a DQN case and a Rainbow case? I think Dueling network still needs to be used in a distributional situation.

Reply:
That's what I mean above, in Rainbow, you don't call (nn::DuelingNetwork)(state) directly. But use the `else` branch instead here. And you need to document it clearly in the docstring, saying that the structure of the `DuelingNetwork` used in Rainbow is different than the conventional one.

Reply:
But if I don't call (nn::DuelingNetwork)(state) directly, other code structures require significant changes.